Content.Shared/Ninja/Systems/NinjaGlovesSystem.cs

```diff
@@ -0,0 +1,325 @@
+    }
+
+    private void EndedDoAfter(EntityUid uid, NinjaGlovesComponent comp, DoAfterEvent args)
+    {
+        comp.Busy = false;
+    }
+
+    /// <summary>
+    /// Helper for glove ability handlers, checks gloves, range, combat mode and stuff.
+    /// </summary>
+    protected bool GloveCheck(EntityUid uid, InteractionAttemptEvent args, [NotNullWhen(true)] out NinjaGlovesComponent? gloves,
+        out EntityUid user, out EntityUid target)
+    {
+        if (args.Target != null && TryComp<NinjaGlovesComponent>(uid, out gloves)
+            && gloves.User != null
+            && !_combatMode.IsInCombatMode(gloves.User)
+            && _timing.IsFirstTimePredicted
+            && TryComp<SharedHandsComponent>(gloves.User, out var hands)
+            && hands.ActiveHandEntity == null)
+        {
+            user = gloves.User.Value;
+            target = args.Target.Value;
+
+            if (_interaction.InRangeUnobstructed(user, target))
+                return true;
+        }
+
+        gloves = null;
+        user = target = EntityUid.Invalid;
+        return false;
+    }
+
+    private void OnDoorjack(EntityUid uid, NinjaDoorjackComponent comp, InteractionAttemptEvent args)
+    {
+        if (!GloveCheck(uid, args, out var gloves, out var user, out var target))
+            return;
+
+        // only allowed to emag non-immune doors
+        if (!HasComp<DoorComponent>(target) || _tags.HasTag(target, comp.EmagImmuneTag))
+            return;
+
+        var handled = _emag.DoEmagEffect(user, target);
+        if (!handled)
+            return;
+
+        ClientPopup(Loc.GetString("ninja-doorjack-success", ("target", Identity.Entity(target, EntityManager))), user, PopupType.Medium);
+        _adminLogger.Add(LogType.Emag, LogImpact.High, $"{ToPrettyString(user):player} doorjacked {ToPrettyString(target):target}");
+    }
+
+    private void OnStun(EntityUid uid, NinjaStunComponent comp, InteractionAttemptEvent args)
+    {
+        if (!GloveCheck(uid, args, out var gloves, out var user, out var target))
+            return;
+
+        // short cooldown to prevent instant stunlocking
+        if (_useDelay.ActiveDelay(uid))
+            return;
+
+        // battery can't be predicted since it's serverside
+        if (user == target || _net.IsClient || !HasComp<StaminaComponent>(target))
+            return;
+
+        // take charge from battery
+        if (!_ninja.TryUseCharge(user, comp.StunCharge))
+        {
+            _popups.PopupEntity(Loc.GetString("ninja-no-power"), user, user);
+            return;
+        }
+
+        // not holding hands with target so insuls don't matter
+        _electrocution.TryDoElectrocution(target, uid, comp.StunDamage, comp.StunTime, false, ignoreInsulation: true);
+        _useDelay.BeginDelay(uid);
+    }
+
+    // can't predict PNBC existing so only done on server.
+    protected virtual void OnDrain(EntityUid uid, NinjaDrainComponent comp, InteractionAttemptEvent args) { }
```

Comment:
```suggestion
    }
    
    private void OnAfterInteract(EntityUid uid, NinjaGlovesComponent component, InteractionAttemptEvent args)
    {
        TryStun(uid, component, args);
        TryDoorJack(uid, component, args);
    }


    private void TryStun(EntityUid uid, NinjaGlovesComponent comp, InteractionAttemptEvent args)
    {
        if (_useDelay.ActiveDelay(uid) || comp.User == args.Target || !HasComp<StaminaComponent>(args.Target) || comp.User == null || _net.IsClient)
            return;

        // take charge from battery
        if (!_ninja.TryUseCharge(comp.User.Value, 36.0f))
        {
            _popups.PopupEntity(Loc.GetString("ninja-no-power"), comp.User.Value, comp.User.Value);
            return;
        }

        // not holding hands with target so insuls don't matter
        _electrocution.TryDoElectrocution(args.Target.Value, uid, 5, TimeSpan.FromSeconds(3), false, ignoreInsulation: true);
        _useDelay.BeginDelay(uid);
    }
    
        private void TryDoorJack(EntityUid uid, NinjaGlovesComponent comp, InteractionAttemptEvent args)
    {
        // only allowed to emag non-immune doors
        if (!HasComp<DoorComponent>(args.Target) || _tags.HasTag(args.Target.Value, "EmagImmune") || comp.User == null)
            return;

        var handled = _emag.DoEmagEffect(comp.User.Value, args.Target.Value);
        if (!handled)
            return;

        ClientPopup(Loc.GetString("ninja-doorjack-success", ("target", Identity.Entity(args.Target.Value, EntityManager))), comp.User.Value, PopupType.Medium);
        _adminLogger.Add(LogType.Emag, LogImpact.High, $"{ToPrettyString(comp.User.Value):player} doorjacked {ToPrettyString(args.Target.Value):target}");
    }
    
        // can't predict PNBC existing so only done on server.
    protected virtual void OnDrain(EntityUid uid, NinjaDrainComponent comp, InteractionAttemptEvent args) { }
```

Reply:
Here's an example on how this would work. Used hardcoded values only for example purposes but this should give the gist of it.

Reply:
would i not just put all abilities in Try* functions then, not just drain and stun